Linspace accepts sympy symbol

timothyman opened the pull request:
- Added K as type for key which can be `str` or `sympy.Symbol`
- Added call to `str` in `SingleSweep` class constructor

Fixes https://github.com/quantumlib/Cirq/issues/1397

timothyman commented:
I signed it!

timothyman commented:
I signed it!

Strilanc commented on cirq/study/sweeps.py:
```diff
@@ -240,11 +240,14 @@ def __str__(self):
 
-    def __init__(self, key: str) -> None:
-        self.key = key
+    def __init__(self, key: K) -> None:
```
Nit: I would just inline `Union[str, sympy.Symbol]` in each case.

Strilanc commented on cirq/study/sweeps.py:
```diff
@@ -240,11 +240,14 @@ def __str__(self):
         return ' + '.join(str(s) for s in self.sweeps)
 
 
+K = TypeVar('K', str, Symbol)  # Key can be of type str or sympy.Symbol
```
You want `Union[str, sympy.Symbol]` instead of a type variable.

timothyman commented:
- Use `import sympy` instead of `from sympy import Symbol` to make `sympy.Symbol` explicit
- Replaced `TypeVar` for `Union[str, sympy.Symbol]`
- Added type-check for `sympy.Symbol` before calling `str` in `SingleSweep` class constructor
